game/hud/src/components/HUD/HUDEditor.tsx

```diff
@@ -0,0 +1,463 @@
+      minScale: 0.5,
+      maxScale: 3,
+      scaleFactor: 0.01,
+    };
+  }
+
+  public render() {
+    const { widgets } = this.props;
+    return (
+      <HUDEditorContainer
+        className='cse-ui-scroller-thumbonly'
+        style={{
+          right: `${this.state.editorPosition.x}px`,
+          top: `${this.state.editorPosition.y}px`,
+        }}>
+        <HUDEditorTitle>
+          <div className='editorDragHandle' onMouseDown={e => this.onMouseDown(e, EditMode.MOVEEDITOR)}>
+            <b>UI Widgets</b>
+            <a href='#' onClick={ () => events.fire('hudnav--navigate', 'reset') }>
+              <div className='resetHUDButton'><b>Reset All</b></div>
+            </a>
+          </div>
+        </HUDEditorTitle>
+        <HUDEditorList>
+          <ul>
+            { _.sortBy(widgets, 'name').map((widget) => {
+              return (widget.name === 'building' ? null : // building should be removed as HUDDrag item
+                <li
+                  key={widget.name}
+                  onClick={() => {
+                    this.props.setSelectedWidget(widget);
+                  }}>
+                  <div className={
+                    this.props.selectedWidget && this.props.selectedWidget.name === widget.name ?
+                      'HUDWidgetNameSelected' : 'HUDWidgetName'
+                  }>
+                    { widget.name === 'motd' ? 'MOTD' : _.startCase(widget.name) }
+                  </div>
+                </li>
+              );
+            }) }
+          </ul>
+        </HUDEditorList>
+        { !this.props.selectedWidget ? <HUDEditorToolbarTitle>- No Widget Selected -</HUDEditorToolbarTitle> : <div>
+          <HUDEditorToolbarTitle>
+            <b>- {this.props.selectedWidget.name === 'motd' ? 'MOTD' : _.startCase(this.props.selectedWidget.name)} -</b>
+          </HUDEditorToolbarTitle>
+          <HUDEditorToolbar>
+
+            { this.props.selectedWidget.widget.dragOptions.lockVisibility ? null :
+              <HUDEditorToolbarItem
+                  onMouseOver={ (e: React.MouseEvent<HTMLElement>) => {
+                    this.tooltipMessage = 'Toggle Visibility';
+                    this.tooltipEvent = e;
+                    this.onMouseOver(e);
+                  } }
+                  onMouseLeave={ this.onMouseLeave }
+              >
+                <a href='#'><i
+                  className={`fa ${this.props.selectedWidget.widget.position.visibility ? 'fa-eye' : 'fa-eye-slash'}`}
+                  onClick={ () => this.setVisibility(
+                    this.props.selectedWidget.name,
+                    !this.props.selectedWidget.widget.position.visibility,
+                  ) }>
+                </i></a>
+              </HUDEditorToolbarItem>
+            }
+
+            { this.props.selectedWidget.widget.dragOptions.lockOpacity ? null :
+              <HUDEditorToolbarItem
+                onMouseOver={ (e: React.MouseEvent<HTMLElement>) => {
```

Comment:
We should declare this logic in a separate function outside of the render function. I see that you're doing it with onMouseOver, onMouseLeave, etc. but we can make more specific functions for these actions that use those other functions.

For example,

```
private onMouseOverAdjustOpacity = (e: React.MouseEvent<HTMLElement>) => {
  this.tooltipMessage = 'Adjust Opacity';
  this.tooltipEvent = e;
  this.onMouseOver(e);
}
```

and then this turns into

`onMouseOver={this.onMouseOverAdjustOpacity}`

It's mainly for readability of the render function. Functions created inside small render functions are not really a problem, but with a render function as big as this one, using more specific functions will help a lot with readability.

Reply:
✔️